File: trunk/src/java/org/codehaus/testdox/intellij/NullPsiElement.java
package org.codehaus.testdox.intellij;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

import com.intellij.psi.PsiElement;

import static jedi.functional.FunctionalPrimitives.array;

public abstract class NullPsiElement implements PsiElement {

    public static final PsiElement INSTANCE;
    static {
        InvocationHandler nullPsiElementInvocationHandler = new InvocationHandler() {
            public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
                return null;
            }
        };

        Class[] interfaces = array(PsiElement.class);
        ClassLoader classLoader = nullPsiElementInvocationHandler.getClass().getClassLoader();

        INSTANCE = (PsiElement) Proxy.newProxyInstance(classLoader, interfaces, nullPsiElementInvocationHandler);
    }

    private NullPsiElement() { }
}

File: src/scala/org/codehaus/testdox/intellij/NullPsiElement.scala
package org.codehaus.testdox.intellij

import com.intellij.psi.PsiElement
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy

object NullPsiElement {

  private val nullPsiElementInvocationHandler = new InvocationHandler() {
    override def invoke(proxy: Object, method: Method, args: Array[Object]): Object = null
  }

  private val interfaces = Array(classOf[PsiElement]).asInstanceOf[Array[Class[_]]]
  private val classLoader = nullPsiElementInvocationHandler.getClass().getClassLoader()

  val INSTANCE = Proxy.newProxyInstance(classLoader, interfaces, nullPsiElementInvocationHandler).asInstanceOf[PsiElement]
}
